### PR TITLE
feat(cluster): Add `--cluster_id` flag

### LINKED ISSUE #2643: persistent node ids in cluster mode

### PR DESCRIPTION
**Pull request by chakaz**
This flag sets the unique ID of a node in a cluster.

It is UB (and bad) to set the same ID to multiple nodes in a cluster.

If unset (default), the `master_replid` (previously known as `master_id`) is used.

Fixes #2643
Related to #2636

### CODE REVIEW

**Review by kostasrim — APPROVED** [review]

**Comment by romange** [comment]
I think cluster id could be read literally as an id of the cluster. I suggest changing to "node_id" because nodes are usually the building blocks in the context of cluster.

**Comment by chakaz** [comment]
How about `--cluster_node_id`?

File: src/server/cluster/cluster_family.cc [hunk 1]
```diff
@@ -25,6 +25,9 @@
 #include "server/server_state.h"
 
 ABSL_FLAG(std::string, cluster_announce_ip, "", "ip that cluster commands announce to the client");
+ABSL_FLAG(std::string, cluster_id, "",
```
Comment (adiholden) [inline src/server/cluster/cluster_family.cc:28]: Did you check if in redis is the node id the same as the cluster node id? 
Lets add it to feature tests of cluster i.e
we should have a test for cluster running with replicas and sending data via seeder (if not we need to add), in this test one of the configurations to run is with this flag

Comment (adiholden) [reply src/server/cluster/cluster_family.cc:28]: and also is there a reason not to set this id for the master_replid_ but have different ids

Comment (romange) [reply src/server/cluster/cluster_family.cc:28]: suppose we always ` master_replid_` to constant `cluster_node_id`, then in case a master restarts, it will still have the same id and the replica will happily replicate empty data from it. that's what we are trying to prevent, right?

Comment (chakaz) [reply src/server/cluster/cluster_family.cc]: > Did you check if in redis is the node id the same as the cluster node id?

No, it's not the same ID. Here's a log from a local redis cluster I set:

```
177592:M 07 Mar 2024 13:01:29.548 * No cluster configuration found, I'm 09a6bad961e78159056f8ad3865a3e2d47d74045
177592:M 07 Mar 2024 13:01:29.560 * Server initialized
[...]
177592:M 07 Mar 2024 13:02:15.440 * Replica 127.0.0.1:6379 asks for synchronization
177592:M 07 Mar 2024 13:02:15.440 * Full resync requested by replica 127.0.0.1:6379
177592:M 07 Mar 2024 13:02:15.440 * Replication backlog created, my new replication IDs are '67e4e538e2e0a76da7a1454615c5ce64954c4800' and '0000000000000000000000000000000000000000'
```

So it's not using the same ID.

> Lets add it to feature tests of cluster i.e we should have a test for cluster running with replicas and sending data via seeder (if not we need to add), in this test one of the configurations to run is with this flag

Done.

File: src/server/server_family.cc [hunk 2]
```diff
@@ -2385,7 +2385,7 @@ void ServerFamily::ReplConf(CmdArgList args, ConnectionContext* cntx) {
         // The response for 'capa dragonfly' is: <masterid> <syncid> <numthreads> <version>
         auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
         rb->StartArray(4);
-        rb->SendSimpleString(master_id_);
+        rb->SendSimpleString(master_replid_);
```
Comment (adiholden) [inline src/server/server_family.cc:2388]: I dont understand how this is the same as cluster node id ..
master_replid_ is set in this line
master_replid_ = GetRandomHex(eng, CONFIG_RUN_ID_SIZE);
I dont see where you that you overide it with cluster node id

File: tests/dragonfly/cluster_test.py [hunk 3]
```diff
@@ -305,8 +316,11 @@ async def test_cluster_slot_ownership_changes(df_local_factory: DflyInstanceFact
 
 
 # Tests that master commands to the replica are applied regardless of slot ownership
+@pytest.mark.parametrize("set_cluster_node_id", [True, False])
 @dfly_args({"proactor_threads": 4, "cluster_mode": "yes"})
-async def test_cluster_replica_sets_non_owned_keys(df_local_factory):
+async def test_cluster_replica_sets_non_owned_keys(
+    df_local_factory: DflyInstanceFactory, set_cluster_node_id: bool
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:322]: you dont use this param in the test

File: tests/dragonfly/cluster_test.py [hunk 4]
```diff
@@ -570,14 +584,20 @@ async def test_cluster_blocking_command(df_server):
     await close_clients(c_master, c_master_admin)
 
 
+@pytest.mark.parametrize("set_cluster_node_id", [True, False])
 @dfly_args({"proactor_threads": 4, "cluster_mode": "yes"})
 async def test_cluster_native_client(
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:589]: where do we run replicaof in this test? I cant find it..

File: src/server/server_family.cc (hunk 2 above)
Comment (chakaz) [reply src/server/server_family.cc:2388]: This PR splits the IDs into 2 separate ones, if `--cluster_node_id` is set.
I.e. the ID used for `master_replid_` is *always* random and can't be overridden, while the ID for cluster has a default value of `master_replid_` but *can* be overridden if set as an argument.

File: tests/dragonfly/cluster_test.py (hunk 3 above)
Comment (chakaz) [reply tests/dragonfly/cluster_test.py:322]: Sorry, leftovers. I decided to use the test below instead (`test_cluster_native_client`)

File: tests/dragonfly/cluster_test.py (hunk 4 above)
Comment (chakaz) [reply tests/dragonfly/cluster_test.py:586]: good catch! :)
I added it, even though it doesn't really matter too much, because we never reach out to the replicas.. should we?

File: src/server/server_family.cc (hunk 2 above)
Comment (adiholden) [reply src/server/server_family.cc:2388]: Ok sorry I miss understood your response on the other thread. So if this is not the same id it as the master_replid_ we should say that we do not support this flag for emulated cluster because in emulated cluster we reply the replica / master  etl.remote_client_id_ which is the master_replid_ generated

File: tests/dragonfly/utility.py [hunk 5]
```diff
@@ -46,12 +46,8 @@ async def wait_available_async(client: aioredis.Redis, timeout=10):
     start = time.time()
     while (time.time() - start) < timeout:
         try:
-            await client.get("key")
+            await client.ping()
```
Comment (adiholden) [inline tests/dragonfly/utility.py:49]: why did you change this?
we want here to make sure that server exits loading stage

File: tests/dragonfly/cluster_test.py (hunk 4 above)
Comment (adiholden) [reply tests/dragonfly/cluster_test.py:586]: We have this in the test.. I guess it is not connecting to the replica as we thought, as the test would have failed before because we did not really replicated
    # Make sure that getting a value from a replica works as well.
    replica_response = await client.execute_command(
        "get", "key0", target_nodes=aioredis.RedisCluster.REPLICAS
    )
    assert "value" in replica_response.values()

Comment (chakaz) [reply tests/dragonfly/cluster_test.py:586]: again, very nice catch!
The reason why it worked before is that it followed `MOVED` error replies to the relevant master, and got the response from it.
Now I use the connections directly, which does not follow `MOVED` (I checked and it fails if I comment out the `replicaof` flag), and now the test works.

**Review by adiholden — APPROVED** [review]
